homeassistant/helpers/script.py

```diff
@@ -180,12 +182,30 @@ def _async_call_service(self, action, variables):
         yield from service.async_call_from_config(
             self.hass, action, True, variables, validate_config=False)
 
-    def _async_fire_event(self, action):
+    def _async_fire_event(self, action, variables):
         """Fire an event."""
         self.last_action = action.get(CONF_ALIAS, action[CONF_EVENT])
         self._log("Executing step %s" % self.last_action)
+        event_data = dict(action.get(CONF_EVENT_DATA, {}))
+        if CONF_EVENT_DATA_TEMPLATE in action:
+            try:
+                def _data_template_creator(value):
```

Comment:
This should be extracted as a method and moved to the template helper: `render(hass, obj)`

Reply:
Because I am a hack (sysadmin, not a professional coder), I had basically just copied this from service.py/async_call_from_config (notice I did add the try/except there since it was annoying that template errors cause exceptions there).

But I very much like the idea of moving this to the template render since it would make it a lot easier to add templating to other functions. I'll give it a whack.

More of a future question, but how opposed would you be to having all data* structures go through templating and not explicitly require data: or data_template:?. The only real downside I can see would be some additional overhead and maybe latency on calls going through the render but it would be fairly minimal I would think. Just a thought, it's not a big deal, but could open up a lot of creative possibilities.

Reply:
The problem with running things through templates is that it ends up as a string.

Reply:
Ah, so things like entity_id are preserved as objects, etc. But aren't they passed as strings anyway if they are included as part of a data_template since they go through the template render process? Don't answer, Lol I don't have time to take enough of a stab so I'll just work on one thing at a time.

Reply:
Looks like this whole construct was for this issue: https://github.com/home-assistant/home-assistant/issues/2985

So this was put in to support lists and dicts as an object that contains template strings, which up until now was only supported in scripts.py (where I stole the code).

But you want to move this functionality into the render() function directly so we resolve the entire list or dict there along with single elements as it currently does, right?

I'll keep hacking at it but wanted to make double sure that was the direction you wanted.

Reply:
No no, we should not move it into an existing function. Instead create a new one that can handle complex data types.

Reply:
I have banged on this for more time than I'd like to admit today and have to admit I am not capable of reworking this. Not saying I couldn't given enough time, but cutting my losses. You can reject this if you want; I'll implement the workaround of passing attributes through a sensor or something

Reply:
@balloob I don't know what notifications you get on these, so figured I would repost. Sorry, was frustrated banging on this in between work and a lack of sleep (which will be you soon, congrats!). So this is implemented as a function in template.py, and cleaned up in script.py and service.py. Updated unit tests for all 3. Looks good to me at this point.